### PR TITLE
fix: Diverse model seeding across PP ranks

### PR DESCRIPTION
**Pull request by rrutmann**
# What does this PR do?

This PR gives a unique model seed for each pp rank, such that parameters are initialized differently across ranks.

## General Changes
* On each rank, add the pp rank to the model seed.

## Breaking Changes
* None

## Checklist before submitting final PR
- [x] My PR is minimal and addresses one issue in isolation
- [x] I have merged the latest version of the target branch into this feature branch
- [x] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [x] I have run a sample config for model training
- [x] I have checked that all tests run through (`python tests/tests.py`)
- [ ] I have updated the internal changelog (`CHANGELOG_DEV.md`)

### CODE REVIEW

**Review by BlueCrescent — COMMENTED** [review]
Overall LGTM.
Should we also explicitly allow seeding for the "model_initialized" component?
It will probably inherit the random state from the model_raw component but it seems a bit risky to me to assume that (also in the future) no other interaction with the random state happens between these two components (though, probably, only interactions that are asymmetrical between the ranks would be problematic). In particular, since we cannot guarantee the order in which the components are build, something like a dataloader component might even re-seed the random state.

**Review by le1nux — CHANGES_REQUESTED** [review]
I checked the seeding (not the test) and from my understanding the changes do not provide the expected results (also what @BlueCrescent  was hinting towards). 

When we seed the raw model, the model weights are indeed deterministic at instantiation time. However, we also have model weight initialization which runs afterwards and would just override the weights / seeding. 

Additionally, passing `device_mesh` to the model is coupling two components that should normally not know anything about each other. 

I think we have to integrate the seeding to the weight initializer component and can also pass in the `device_mesh` there.

**Comment by rrutmann** [comment]
> I checked the seeding (not the test) and from my understanding the changes do not provide the expected results (also what @BlueCrescent was hinting towards).
> 
> When we seed the raw model, the model weights are indeed deterministic at instantiation time. However, we also have model weight initialization which runs afterwards and would just override the weights / seeding.
> 
> Additionally, passing `device_mesh` to the model is coupling two components that should normally not know anything about each other.
> 
> I think we have to integrate the seeding to the weight initializer component and can also pass in the `device_mesh` there.

Yes that makes sense. I moved the seeding to the model initialization component

**Comment by rrutmann** [comment]
> Overall LGTM. Should we also explicitly allow seeding for the "model_initialized" component? It will probably inherit the random state from the model_raw component but it seems a bit risky to me to assume that (also in the future) no other interaction with the random state happens between these two components (though, probably, only interactions that are asymmetrical between the ranks would be problematic). In particular, since we cannot guarantee the order in which the components are build, something like a dataloader component might even re-seed the random state.

See https://github.com/Modalities/modalities/pull/426#issuecomment-3675521456

File: src/modalities/models/model.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ class ActivationType(str, Enum):
 class NNModel(nn.Module):
     """NNModel class to define a base model."""
 
-    def __init__(self, seed: int = None, weight_decay_groups: Optional[WeightDecayGroups] = None):
+    def __init__(self, seed: Optional[int] = None, weight_decay_groups: Optional[WeightDecayGroups] = None):
```
Comment (le1nux) [inline src/modalities/models/model.py]:
```suggestion
    def __init__(self, seed: int | None = None, weight_decay_groups: Optional[WeightDecayGroups] = None):
```

Comment (le1nux) [reply src/modalities/models/model.py]: Do we even want to allow setting the seed here? 
Could `torch.manual_seed` below have side effects with the new weight init implementation?

File: src/modalities/nn/model_initialization/composed_initialization.py [hunk 2]
```diff
@@ -30,6 +32,8 @@ class ComposedModelInitializationConfig(BaseModel):
     std: Annotated[float, Field(strict=True, ge=0.0)] | str  # can be float or "auto"
     hidden_dim: Optional[Annotated[int, Field(strict=True, gt=0)]] = None
     num_layers: Optional[Annotated[int, Field(strict=True, gt=0)]] = None
+    seed: Optional[int] = None
```
Comment (le1nux) [inline src/modalities/nn/model_initialization/composed_initialization.py]:
```suggestion
    seed: int | None = None
```

Comment (le1nux) [reply src/modalities/nn/model_initialization/composed_initialization.py]: also other instances and docstring annotations should be adapted accordingly

File: src/modalities/nn/model_initialization/initialization_routines.py [hunk 3]
```diff
@@ -109,20 +119,29 @@ def get_scaled_initialization(
...
         """
+        InitializationRoutines._set_seed(seed)
         std = math.sqrt(0.4)
         initialization = NamedParameterwiseNormalInitialization(
             mean=mean, std=std, parameter_name_regexes=parameter_name_regexes
         )
         return initialization
+
+    @staticmethod
+    def _set_seed(seed: Optional[int]):
```
Comment (le1nux) [inline src/modalities/nn/model_initialization/initialization_routines.py]: this sets the seed globally. I think an even more robust way would be to use a local torch rng object. 
Could this be integrated? 

Something like: 
```py
g = torch.Generator()
g.manual_seed(1234)
```

**Review by BlueCrescent — APPROVED** [review]
LGTM

File: src/modalities/nn/model_initialization/composed_initialization.py [hunk 4]
```diff
@@ -114,16 +120,28 @@ def get_composed_model_initializer(
...
+            seed (Optional[int], optional): Seed for random initialization. Defaults to None.
 
         Returns:
             ModelInitializationIF: The Weight Initializer performing the initialization as specified.
         """
+        # Set different random seed for each PP rank to ensure diversity
+        if seed is not None and has_parallelism_method(
+            device_mesh=device_mesh, parallelism_method=ParallelismDegrees.PP
+        ):
+            seed += get_parallel_rank(device_mesh=device_mesh, parallelism_method=ParallelismDegrees.PP)
```
Comment (le1nux) [inline src/modalities/nn/model_initialization/composed_initialization.py:169]: This also means that depending on the parallelization method and also the number of parallelism degrees we get differently initialized layers even if the seed is the same.
Example: 

DP with seed = 1, will have a differently initilized model than DP+PP with seed = 1. 

One way to fix this is to always use the same seed but each PP stage has to skip the number of random values of the pervious stages. 
However, I think this would be overkill and I would just place a warning when initialising the weights and parallelization methods are other than FSDP.
What do you think?

Comment (le1nux) [reply src/modalities/nn/model_initialization/composed_initialization.py:169]: We could also give the hint that for full reproducibility a Distributed Checkpoint with FSDP directly after weight init. 
Maybe we could even have an entry point for that in main. 

something like:
`modalities create_init_cp model_config.yaml`

For some unit tests, this functionality would be nice to have anyways I think.

Any thoughts?

File: src/modalities/nn/model_initialization/initialization_routines.py [hunk 5]
```diff
@@ -62,7 +63,11 @@ def initialize_in_place(self, model: nn.Module):
 class InitializationRoutines:
     @staticmethod
     def get_plain_initialization(
-        mean: float, std: float | str, parameter_name_regexes: list[str], hidden_dim: Optional[int] = None
+        mean: float,
+        std: float | str,
+        parameter_name_regexes: list[str],
+        hidden_dim: Optional[int] = None,
+        seed: Optional[int] = None,
```
Comment (le1nux) [inline src/modalities/nn/model_initialization/initialization_routines.py]:
```suggestion
        seed: int | None = None,
```

also various cases below

File: src/modalities/nn/model_initialization/initialization_routines.py (hunk 3 above)
Comment (le1nux) [reply src/modalities/nn/model_initialization/initialization_routines.py]: Otherwise, we might get into side-effects later on

**Review by le1nux — CHANGES_REQUESTED** [review]
Generally good state. 
Left a couple of comments. 
My main concern is the global setting of the seed. A generator object might be favorable.

File: src/modalities/models/model.py (hunk 1 above)
Comment (rrutmann) [reply src/modalities/models/model.py]: Probably it could have side effects, e.g. default submodule initialization, random ops and the ambient global RNG state for unrelated code. Also it is mostly redundant since we now use a local generator for weight initialization. I would suggest to remove it.

File: src/modalities/nn/model_initialization/initialization_routines.py (hunk 3 above)
Comment (rrutmann) [reply src/modalities/nn/model_initialization/initialization_routines.py]: Makes sense. I integrated this

File: src/modalities/nn/model_initialization/composed_initialization.py (hunk 4 above)
Comment (rrutmann) [reply src/modalities/nn/model_initialization/composed_initialization.py:169]: Good catch. I added a warning for now, but the additional entry point would be nice to have as well. I created an [issue](https://github.com/Modalities/modalities/issues/443) for that

File: src/modalities/conversion/gpt2/modeling_gpt2.py [hunk 6]
```diff
@@ -40,7 +40,14 @@
 from transformers.modeling_utils import ALL_ATTENTION_FUNCTIONS, PreTrainedModel
 from transformers.processing_utils import Unpack
 from transformers.utils import TransformersKwargs, auto_docstring, can_return_tuple, logging
-from transformers.utils.generic import check_model_inputs
+
+try:
+    from transformers.utils.generic import check_model_inputs
+except ImportError:
+
+    def check_model_inputs(func: Callable) -> Callable:
```
Comment (le1nux) [inline src/modalities/conversion/gpt2/modeling_gpt2.py:48]: was this removed in transormers? 
If it is part of a legacy API I think we should also remove this on our end. 
What do you think @BlueCrescent? I think you added it, right?

File: src/modalities/nn/model_initialization/initialization_routines.py [hunk 7]
```diff
@@ -39,21 +40,32 @@ class ScaledEmbedInitializationConfig(BaseModel):
...
 class NamedParameterwiseNormalInitialization(ModelInitializationIF):
-    def __init__(self, mean: float, std: float, parameter_name_regexes: RegexFilter):
+    def __init__(self, mean: float, std: float, parameter_name_regexes: RegexFilter, seed: int | None = None):
         self.mean = mean
         self.std = std
         self.parameter_name_regexes = parameter_name_regexes
+        self.seed = torch.initial_seed() if seed is None else seed
+        self._generators: dict[str, torch.Generator] = {}
+
+    def _get_generator(self, parameter: torch.Tensor) -> torch.Generator:
```
Comment (le1nux) [inline src/modalities/nn/model_initialization/initialization_routines.py:66]: a few things are not clear to me. 
 1. Do we actually have the case, where in a single process tensors are sitting on different GPUs? 
 2. if 1. is the case, then we can end up with tensors that are initialized identically, since we create multiple generators from the same seed. 

I'm not sure what the best way to solve this ... also seems to me that the Pytorch API regarding Generators is kinda limited.

File: src/modalities/nn/model_initialization/initialization_routines.py [hunk 8]
```diff
@@ -73,23 +89,26 @@ def get_plain_initialization(
             std (float): standard deviation of the normal distribution. If set to "auto", appropiate
                 value selected as per plain initialization described in https://arxiv.org/abs/2312.16903
             hidden_dim (Optional[int]): hidden dimension of the attention layer. Defaults to None.
+            parameter_name_regexes (list[str]): List of parameter name regexes to which the initialization
```
Comment (le1nux) [inline src/modalities/nn/model_initialization/initialization_routines.py:118]: I think the type annotation is wrong. should be `RegexFilter`

File: src/modalities/nn/model_initialization/initialization_routines.py [hunk 9]
```diff
@@ -99,6 +118,7 @@ def get_scaled_initialization(
             num_layers (int): Number of layers in the model which we use to downscale std with
             parameter_name_regexes (list[str]): List of parameter name regexes to which the initialization
```
Comment (le1nux) [inline src/modalities/nn/model_initialization/initialization_routines.py]:
```suggestion
            parameter_name_regexes (RegexFilter): List of parameter name regexes to which the initialization
```

**Review by le1nux — CHANGES_REQUESTED** [review]
Looks good to me! Nice work :) 
The comment regarding the per-device Generator we should discuss, what makes most sense here. 

I would add one last test, which checks that two models instantiated with the same config file (with a specified seed), should have 100% matching parameter weights. I'd keep that one simple (no advanced sharding like TP or PP. only FSDP).

File: src/modalities/conversion/gpt2/modeling_gpt2.py (hunk 6 above)
Comment (rrutmann) [reply src/modalities/conversion/gpt2/modeling_gpt2.py:48]: The function was removed in transformers version 5.2. In our pyproject.yaml we specify the requirement  "transformers>=4.57.4,<5.0.0", so I used an unsupported transformers version here. Should we remove it just to be on the safe side?

File: src/modalities/nn/model_initialization/initialization_routines.py (hunk 7 above)
Comment (rrutmann) [reply src/modalities/nn/model_initialization/initialization_routines.py:66]: Since we start a single process for each rank via torchrun, this shouldn't happen, right? Or do I miss something?

File: src/modalities/conversion/gpt2/modeling_gpt2.py (hunk 6 above)
Comment (BlueCrescent) [reply src/modalities/conversion/gpt2/modeling_gpt2.py:48]: Yes, I think, we should tackle the transformers 5.0.0+ support soon anyways.

File: src/modalities/nn/model_initialization/initialization_routines.py [hunk 10]
```diff
@@ -39,21 +41,49 @@ class ScaledEmbedInitializationConfig(BaseModel):
...
 
 class NamedParameterwiseNormalInitialization(ModelInitializationIF):
-    def __init__(self, mean: float, std: float, parameter_name_regexes: RegexFilter):
+    def __init__(
+        self,
+        mean: float,
+        std: float,
+        parameter_name_regexes: RegexFilter,
+        seed: int | None = None,
+        multi_device_generator_policy: Literal["ignore", "warn", "error"] = "warn",
```
Comment (le1nux) [inline src/modalities/nn/model_initialization/initialization_routines.py]: Since we instantiate this from outside, I would make this an enum

**Review by le1nux — APPROVED** [review]
Minor comment. Otherwise all looks good to me! :) 
Nice work! 👍
